plugins/inputs/elasticsearch_query/elasticsearch_query.go

```diff
@@ -0,0 +1,304 @@
+  ## Must be text, non-analyzed fields. Metric aggregations are performed per tag
+  # tags = ["field.keyword", "field2.keyword"]
+
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  # include_missing_tag = false
+
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  # missing_tag_value = "null"
+`
+
+// ElasticsearchQuery struct
+type ElasticsearchQuery struct {
+	URLs                []string          `toml:"urls"`
+	Username            string            `toml:"username"`
+	Password            string            `toml:"password"`
+	EnableSniffer       bool              `toml:"enable_sniffer"`
+	Timeout             internal.Duration `toml:"timeout"`
+	HealthCheckInterval internal.Duration `toml:"health_check_interval"`
+	Aggregations        []esAggregation   `toml:"aggregation"`
+	tls.ClientConfig
+	httpclient *http.Client
+	esClient   *elastic5.Client
+}
+
+// esAggregation struct
+type esAggregation struct {
+	Index                string            `toml:"index"`
+	MeasurementName      string            `toml:"measurement_name"`
+	DateField            string            `toml:"date_field"`
+	QueryPeriod          internal.Duration `toml:"query_period"`
+	FilterQuery          string            `toml:"filter_query"`
+	MetricFields         []string          `toml:"metric_fields"`
+	MetricFunction       string            `toml:"metric_function"`
+	Tags                 []string          `toml:"tags"`
+	IncludeMissingTag    bool              `toml:"include_missing_tag"`
+	MissingTagValue      string            `toml:"missing_tag_value"`
+	mapMetricFields      map[string]string
+	aggregationQueryList []aggregationQueryData
+}
+
+// SampleConfig returns sample configuration for this plugin.
+func (e *ElasticsearchQuery) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns the plugin description.
+func (e *ElasticsearchQuery) Description() string {
+	return `Derive metrics from aggregating Elasticsearch query results`
+}
+
+// Init the plugin.
+func (e *ElasticsearchQuery) Init() error {
+	if e.URLs == nil {
+		return fmt.Errorf("elasticsearch urls is not defined")
+	}
+
+	err := e.connectToES()
+	if err != nil {
+		log.Printf("E! error connecting to elasticsearch: %s", err)
+		return nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), e.Timeout.Duration)
+	defer cancel()
+
+	for i, agg := range e.Aggregations {
+		if agg.MeasurementName == "" {
+			return fmt.Errorf("field 'measurement_name' is not set")
+		}
+		if agg.DateField == "" {
+			return fmt.Errorf("field 'date_field' is not set")
+		}
+		e.initAggregation(ctx, agg, i)
+	}
+	return nil
+}
+
+func (e *ElasticsearchQuery) initAggregation(ctx context.Context, agg esAggregation, i int) {
+	// retrieve field mapping and build queries only once
+	mapMetricFields, err := e.getMetricFields(ctx, agg)
+	if err != nil {
+		log.Printf("E! %s", err.Error())
+		return
```

Comment:
Wouldn't it make sense to return this error and handle it in `Init()`? Can you do some sensible processing if this fails?

Reply:
The idea here is just to wait for the next plugin execution as the data for the query may not be available (yet) on ES, and the plugin init should not fail because of that.

Reply:
Fair enough.